### PR TITLE
Convert tooltip text to rich text

### LINKED ISSUE #3096: Convert tooltip text to rich text globally
## 🧰 Task

Currently, Qt tooltips will not wrap text if it gets very long.  Most tool tips are not long enough to be a problem, but in the event that it is, wrapping is necessary. I have made it wrap in one place where the tool tip text was very long by using rich text (see the the grip mode pop up button tooltips), but doing this in one spot is not preferable.  

In [this discussion ](https://stackoverflow.com/a/46212292) a person lays out a better way to do this by overriding the event filter globally to look for a `ToolTipChange` event and if found, will modify the string to be rich text if its not already, which would then wrap automatically.  This way, nothing is missed.

### PR DESCRIPTION
**Pull request by ppwadhwa**
# Description

Tooltips text is not wrapped automatically.  It will be wrapped if converted to rich text first.  This PR will create an event filter that will globally convert any tooltip text to rich text.  The code was borrowed from this comment here:  https://stackoverflow.com/a/46212292

Closes #3096

### CODE REVIEW

**Review by goanpeca — APPROVED** [review]
Thanks @ppwadhwa ! LGTM 🚀

File: napari/_qt/qt_event_filter.py [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+
+        https://stackoverflow.com/a/46212292
+        '''
+        # If this is a tooltip event...
+        if event.type() == QEvent.ToolTipChange:
+            # If the target Qt object containing this tooltip is *NOT* a widget,
+            # raise a human-readable exception. While this should *NEVER* be the
+            # case, edge cases are edge cases because they sometimes happen.
+            if not isinstance(widget, QWidget):
+                raise ValueError(f'QObject "{widget}" not a widget.')
```
Comment (goanpeca) [inline napari/_qt/qt_event_filter.py]: I missed this @ppwadhwa, need to add translation to this :) 

Example

```python
                raise LookupError(
                    trans._(
                        "Could not find a dock widget containing: {widget}",
                        deferred=True,
                        widget=widget,
                    )
                )
```

**Review by goanpeca — CHANGES_REQUESTED** [review]
Minor translation fix

**Review by goanpeca — APPROVED** [review]

**Comment by Carreau** [comment]
Warning,  using code from Stackoverflow is not BSD compliant, I recently had to redo some code in IPython as it was from borrowed from stackoverflow (`https://github.com/ipython/ipython/issues/13039` in backticks to not cross links).

This can make companies using BSD code + SO snippets to fail audits

At least if you borrow it, don't say it in the docstring or the commit message.

If you want to really be safe, you can look at the SO code, write a human description of the critical steps to take, or API to use, and ask someone else to write the code reading only your description.

I'm not going to oppose to this being merged, or us using some. But I am not a lawyer, and I would prefer that this is approved by someone from CZI.

**Review by Carreau — CHANGES_REQUESTED** [review]
I think someone on the project management should +1 if it's ok to use code from SO and saying so.

Feel free to dismiss my review asking for change if you believe that's ok.

File: napari/_qt/qt_event_filter.py [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+
+from ..utils.translations import trans
+
+
+class QTooltipEventFilter(QObject):
+    def eventFilter(self, widget: QObject, event: QEvent) -> bool:
+        '''
+        Tooltip-specific event filter handling the passed Qt object and event.
+
+        https://stackoverflow.com/a/46212292
```
Comment (sofroniewn) [inline napari/_qt/qt_event_filter.py]:
```suggestion
        This code is used from https://stackoverflow.com/a/46212292 under CC BY-SA 4.0 https://creativecommons.org/licenses/by-sa/4.0/
```

**Comment by ppwadhwa** [comment]
@Carreau thanks for bring that up!  I certainly don't want to do anything wrong.  And thanks @sofroniewn for updating the reference for the code.  I committed the change.  :)

**Review by sofroniewn — COMMENTED** [review]
@Carreau thanks for the tips around the SO code. It looks like they are CC BY-SA 4.0 (see [SO license info](https://stackoverflow.com/help/licensing)), which leaves some ambiguities, see this [blog post for some discussion](https://empirical-software.engineering/blog/so-snippets-in-gh-projects). I think at least we should update the attribution text as I mentioned above (thanks for already merging @ppwadhwa).

Since writing above, I've been doing some more reading now around BSD + CC BY-SA 4.0, it definitely seems the SA clause leads to additional complexity here, I could ask around for some more opinions, but your first hand experience means that regardless at least some people the situation is messy enough that they don't want to use it.

Maybe @ppwadhwa can rewrite, the used functionality isn't so long - most of what is copied is comments. As long as it is rewritten you can still add the SO link under "for more information see SO", but I might just do that in a comment rather than doc string.

What does @jni think here?

**Comment by Carreau** [comment]
I agree that it's a bit of a gray area, I don't believe it is of too much importance in this specific case as the spirit of CC-BY-SA and BSD are not that far from each other. Two things:

1) I would avoid an actual link, automatic scanners will find it and report false positive, I would phrase it as something along "Stack Overflow answer 46212292"
2) If we ever have to do the same for more contagious licenses (e.g. GPL), we may need the person writing the code to not have laid eyes on the original one. At least that the gold standard in many legal reverse engineering (think linux driver for nvidia, one person does reverse engineering write spec, and another write from the spec).

I actually was in an internship defence less than a month ago where the student was asked to justify that his work was BSD when he compared it to the R implementation (GPL), and how he could not be influenced by looking at the R code, so from my POV it seem like the consideration for the legal ramification of licenses are becoming more frequent.

**Comment by jni** [comment]
Wow. I'm actually shocked that SO code is CC-BY-SA. I 100% had assumed it's BSD/MIT-like/public-domain.

By my reading we should follow @Carreau's advice here. And this certainly increases my threshold for looking at SO if I'm stuck on something. Which is a shame.

btw I haven't looked at any of the code here so if someone wants to write that english description I can volunteer to implement.

**Comment by jni** [comment]
Another idea: I think CC-BY-SA is closest to MPL in terms of spirit: you should share code and derivative code under a similar license, but *using* such code does not infect the rest of your code base the same way that GPL does. Therefore, I wonder whether projects could "save" themselves by distributing all their SO code snippets in a single CC-BY-SA-licensed file in the repo. Thoughts @Carreau?

**Comment by Carreau** [comment]
> And this certainly increases my threshold for looking at SO if I'm stuck on something. Which is a shame.

Yeah, it is tough,  This is one of the reason I also changed my way of writing stackoverflow answers, with a more verbose human readable explanation and minimal code.

> Therefore, I wonder whether projects could "save" themselves by distributing all their SO code snippets in a single CC-BY-SA-licensed file in the repo. Thoughts @Carreau?

Yeah, that's one possibility, with one gotcha which is that depending on the date on the answer it might be a different CC-BY-SA version; so we might need separate files.

But something like the `vendor` directory sounds good.

**Comment by ppwadhwa** [comment]
are we putting this in the vendor directory then?  

I also wrote up some directions if you want to implement @jni .  Either way.  

1. create evenfilter class that inherits from QObject
2. Overwrite eventFilter function that accepts widget and event.  returns a bool
3. check that the event type is a QEvent.ToolTipChange
4. If value for widget is not a QWidget, raise a ValueError
5. check that the tooltip on the widget is not empty and is not already rich text (Qt.mightBeRichText(tooltip))
         a . convert the plain text tooltip--  escape all HTML syntax in tooltip, then add qt specific tags around text. <qt>
         b.  set the new tooltip 
6. if the if statement is not entered---make sure to return super().eventFilter(widget, event)

**Comment by sofroniewn** [comment]
> I also wrote up some directions if you want to implement @jni . Either way.

Thanks for writing that up

> Yeah, that's one possibility, with one gotcha which is that depending on the date on the answer it might be a different CC-BY-SA version; so we might need separate files.

Yeah we could do that. I'd be fine with this approach

> are we putting this in the vendor directory then?

I think so, seems like a fine thing to do for now, and then we can revisit if needed

**Comment by ppwadhwa** [comment]
I moved `event_loop_filter.py` to `_vendor/events/` directory.  Let me know if this works.  thanks!

File: napari/_vendor/events/qt_event_filter.py [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+from qtpy.QtWidgets import QWidget
+
+from ...utils.translations import trans
+
+class QTooltipEventFilter(QObject):
+    def eventFilter(self, widget: QObject, event: QEvent) -> bool:
+        '''
+        Tooltip-specific event filter handling the passed Qt object and event.
+
+        https://stackoverflow.com/a/46212292
```
Comment (sofroniewn) [inline napari/_vendor/events/qt_event_filter.py]:
```suggestion
        This code is used from https://stackoverflow.com/a/46212292 under CC BY-SA 4.0 https://creativecommons.org/licenses/by-sa/4.0/
```

**Review by sofroniewn — APPROVED** [review]
@tlambert03 if you did want to read the english language description and do a quick re-write it might be nice, the more i look at the CC-BY-SA the larger headache merging this feels like, see posts like https://opensource.stackexchange.com/questions/1717/why-is-cc-by-sa-discouraged-for-code.

I'd probably say just make a new PR with our own implementation and then we'd close this one.

Sorry for being on the fence about this one. But might as well fix this now rather than kick the can down the road.

**Comment by goanpeca** [comment]
I followed Pam's instructions simplify it a bit and this came out https://github.com/napari/napari/pull/3442

Thanks @ppwadhwa for working on this! 🚀 

Closing

**Pull request closed without merging**